README
CheckDigits v1.3.4
==================

This library provides several modules to compute or validate check
digits.

There is a executable script that may be used at the command line to
compute or validate check digits.

Further there is a sample CGI script that may be used to setup a web
service to validate check digits in your webbrowser.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires no other modules and libraries.


COPYRIGHT AND LICENCE

Copyright (C) 2003-2020 Mathias Weidner

This module may be distributed under the same license as perl itself.


Changes
Revision history for Perl extension CheckDigits.

v1.3.4 Di 17. Nov 15:13:54 CET 2020
	- revise lib/Algorithm/CheckDigits/M23_002.pm (irish TIN)
	- fix Bug #133031

v1.3.3 Di 14. Jul 19:20:56 CEST 2020
	- reworked lib/Algorithm/CheckDigits.pm, allow 10-digit numbers,
	  allow prepending of 'BE' before the digits
	- fix Bug #132950

v1.3.2 Mo 16. Mai 15:02:03 CEST 2016
	- fix version strings

1.3.1 Mo 16. Mai 03:19:57 CEST 2016
	- fix Bug #114347
	- fix version numbers
	- require Perl version in Build.PL

1.3.0 Sun May 12 10:28:21 CEST 2013
	- Changed description of modules to more meaningful text (at least I
	  hope so).
	- Added plugin interface with examples in t/PluginLib*.pm and
	  t/plugin*.t.

1.2.1 Sun Mar  3 22:51:45 CET 2013
	- changed list items in POD from 0, 1, ... to S<0>, S<1>, ...
	  according to a thread on the perl5 porters list
	  (www.nntp.perl.org/group/perl.perl5.porters/2013/02/msg199355.html)

	  David Wheeler: "Lists must start with 1. When it saw the 0, it
	  assumed it was a definition list, so the numbers after that
	  don't look right."

1.2.0 Fr 8. Jun 11:19:25 CEST 2012
	- reworked Algorithm::CheckDigits::MXX_001.pm, added ABA routing
	  numbers

1.1.2 Fr 8. Jun 09:22:35 CEST 2012
	- added '=encoding iso-8859-1' to M11_004.pm, M11_006.pm,
	  MBase_002.pm, MXX_003.pm, MXX_006.pm. This should prevent
	  failures in the processing of the POD that contain non-ASCII
	  characters.
	- corrected search and replacement lists in M10_011.pm

1.1.1 Do 3. Mar 10:16:59 CET 2011
	- Reworked Algorithm::CheckDigits::M10_008 (sedol) to cop with
	  alphanumerical SEDOL codes that are in use since March 2004.
	- There was also a flaw in the computation of the checksum that
	  provided 10 instead of 0.
	- This closes bug number #66340 on CPAN.

1.1.0 Fr 12. Nov 19:10:39 CET 2010
	- reworked Algorithm::CheckDigits::M10_001, added CUSIP
	- moved scripts/checkdigits.pl to bin/checkdigits.pl
	- added cgi-bin/checkdigits.cgi

1.0.0 Mo 13. Sep 21:27:23 CEST 2010
	- reworked Algorithm::CheckDigits::MXX_001

0.90.0 Fr 10. Sep 15:52:51 CEST 2010
	- added scripts/checkdigits.pl as sample script

0.53  Fr 20. Nov 11:46:08 CET 2009
	- cleaned up documentation
	  (thanks to Romuald Żyłła for pointing me to more errors)
	- added $VERSION into all modules to close #36482 at
	  rt.cpan.org

0.52  Mi 11. Nov 10:31:18 CET 2009
       - added link to english description for ustid_pl

0.51  Mi 11. Nov 10:01:36 CET 2009
       - corrected documentation for ustid_pl/vatnr_pl
	 (thanks to Romuald Żyłła for pointing me to the error)
       - reworked algorithm and documentation for esr5_ch according to
         http://www.sic.ch/de/dl_tkicch_dta.pdf
	 (thanks to Romuald Żyłła for pointing me to the error)

0.50  Fr 6. Jun 21:45:04 CEST 2008
       - put $VERSION back into CheckDigits.pm

0.49  Fr 16. Mai 22:54:43 CEST 2008
       - CheckDigits/MXX_002.pm computes checkdigits for CAS registration
         numbers with ten digits.

0.48  Di 9. Okt 01:07:02 CEST 2007
	- keep VERSION in Makefile.PL

0.47  Di 9. Okt 00:45:07 CEST 2007
	- provide Version in file VERSION
	- Makefile.PL: indicate minimum perl version
	  (thanks to Slaven Rezic)

0.46  Mo 8. Okt 23:46:32 CEST 2007
	- CheckDigits/M11_017.pm computes EC-No (EINECS, ELINCS) check digits.

0.45  So 7. Okt 12:00:25 CEST 2007
	- CheckDigits.pm: CheckDigits() dies on unknown algorithm/method
	- CheckDigits.pm: IMEI and IMEISV are recognized (M10_001.pm)
	  (Petri Oksanen made me aware that this wasn't so before)
	- CheckDigits/M10_001.pm recognizes that with IMEISV only the 14 most
	  significant digits are taken for the computation of the checkdigit
	- reformatted CheckDigits.pm, CheckDigits/M10_001.pm

0.44  Mo Dez 11 00:46:30 CET 2006
	- Method 97-002 (IBAN): allow min. 2, max. 30 digits/letters for BBAN

0.43  So Dez 10 23:35:31 CET 2006
	- Method 97-002 (IBAN): allowed for more letters than just the country
	  code and eliminated Math::BigInt (Thanks to Detlef Pilzecker).

0.42  Fr Dez  1 22:02:57 CET 2006
	- Method 10-004: added code to handle ISBN-13 (mainly check whether
	  first three digits are 978 or 979 and cope with hyphens)
	- CheckDigits.pm: Invocation as CheckDigits('isbn13') returns an
	  Algorithm::CheckDigits::M10_004 object.
	- added tests for isbn13

0.41  Do Jun 29 00:34:04 CEST 2006
	- added t/pod-coverage.t
	- added missing documentation

0.40  Mi Jun 28 23:31:00 CEST 2006
	- fixed code examples in POD according to
	  http://annocpan.org/~MAMAWE/Algorithm-CheckDigits-0.39/CheckDigits/MXX_002.pm
	- CheckDigits.pm: POD: corrected ABSTRACT
	- added t/pod.t to check for POD errors
	- fixed POD according to t/pod.t
	- Clarified POD for Algorithm/CheckDigits/MXX_006.pm

0.39  Sun, 12 Mar 2006 17:50:13 +0100
	- Added Verhoeff scheme (Method XX-006) taking some code from
	  Algorithm::Verhoeff.
	- Refined POD for CheckDigits.pm

0.38  Wed, 26 Jan 2005 22:48:10 +0100
	- Refined Method M11-011 (now accepts full vatrn)
	- provided basic tests for above change

0.37  Wed, 14 Jul 2004 20:38:31 +0200
	- Method 10-001: Added Support for ISIN
	- Method XX-004 (ustid_at) implemented
	- Method XX-005 (esr9_ch) implemented

0.36  Tue, 13 Jul 2004 23:45:02 +0200
	- Method XX-003 (dem) implemented
	- Method XX-002 (cas) implemented

0.35  Sun, 11 Jul 2004 18:59:35 +0200
	- Method XX-001 (pa_de) implemented

0.34a Fri, 09 Jul 2004 23:57:37 +0200
	- Method 97-002 (iban) implemented

0.34  Wed, 14 Jun 2004 22:04:38 +0200
	- Method 97-001 (ustid_be) implemented

0.33  Wed, 09 Jun 2004 22:04:38 +0200
	- Method 23-002 (ustid_ie) implemented
	- Method 43-001 (code_39) implemented
	- Method 89-001 (ustid_lu) implemented

0.32  Fri, 14 May 2004 17:22:15 +0200
	- Aaron W. West made me aware of a fault in MBase_001.pm (the
	  code providing UPC check digits) and sent me a fix
	- added test routine for that failure

0.31  Tue, 11 May 2004 00:26:56 +0200
	- Method 23-001 (dni_es) implemented
	- Renamed methods 001..010

0.30  Mon, 10 May 2004 01:00:14 +0200

lib/Algorithm/CheckDigits/M10_001.pm
# vim: set ts=4 sw=4 tw=78 si et:
package Algorithm::CheckDigits::M10_001;

use 5.006;
use strict;
use warnings;
use integer;

use version; our $VERSION = qv('v1.3.4');

our @ISA = qw(Algorithm::CheckDigits);

my %prefix = (
    'amex'     => [ '34', '37', ],
    'bahncard' => [ '70', ],
    'diners'   => [ '30[0-5]', '36', '38', ],
    'discover' => [ '6011', ],
    'enroute' => [ '2014', '2149', ],
    'jcb'     => [ '1800', '2131', '3088', ],
    'mastercard' => [ '5[1-5]', ],
    'miles&more' => [ '99', '22', ],
    'visa'       => [ '4', ],
);

my %ctable = (
    '0' => 0, '1' => 1, '2' => 2, '3' => 3, '4' => 4,
    '5' => 5, '6' => 6, '7' => 7, '8' => 8, '9' => 9,
    'A' => 10, 'B' => 11, 'C' => 12, 'D' => 13, 'E' => 14,
    'F' => 15, 'G' => 16, 'H' => 17, 'I' => 18, 'J' => 19,
    'K' => 20, 'L' => 21, 'M' => 22, 'N' => 23, 'O' => 24,
    'P' => 25, 'Q' => 26, 'R' => 27, 'S' => 28, 'T' => 29,
    'U' => 30, 'V' => 31, 'W' => 32, 'X' => 33, 'Y' => 34,
    'Z' => 35,
);

# Aliases
$prefix{'eurocard'} = $prefix{'mastercard'};

# omit prefixes doesn't work with the test numbers
my %omitprefix = (
    'jcb'      => 0,
    'enroute'  => 0,
    'discover' => 0,
);

sub new {
    my $proto = shift;
    my $type  = shift;
    my $class = ref($proto) || $proto;
    my $self  = bless( {}, $class );
    $self->{type} = lc($type);
    $self->_determine_pattern();
    return $self;
}    # new()

sub is_valid {
    my ( $self, $number ) = @_;
    if ( $number =~ /^($self->{pattern})([0-9])$/i ) {
        return $2 == $self->_compute_checkdigit( uc($1) );
    }
    return '';
}    # is_valid()

sub complete {
    my ( $self, $number ) = @_;
    if ( $number =~ /^$self->{pattern}$/i ) {
        return $number . $self->_compute_checkdigit( uc($number) );
    }
    return '';
}    # complete()

sub basenumber {
    my ( $self, $number ) = @_;
    if ( $number =~ /^($self->{pattern})([0-9])$/i ) {
        return $1 if ( $2 == $self->_compute_checkdigit( uc($1) ) );
    }
    return '';
}    # basenumber()

sub checkdigit {
    my ( $self, $number ) = @_;
    if ( $number =~ /^($self->{pattern})([0-9])$/i ) {
        return $2 if ( $2 == $self->_compute_checkdigit( uc($1) ) );
    }
    return '';
}    # checkdigit()

sub _compute_checkdigit {
    my $self   = shift;
    my $number = shift;
    $number =~ s/\s//g;
    if ( $omitprefix{ $self->{type} } ) {
        my $pf = $prefix{ $self->{type} };
        for my $p ( @{$pf} ) {
            if ( $number =~ /^$p([0-9]+)$/ ) {
                $number = $1;
                last;
            }
        }
    }
    if ('isin' eq $self->{type}) {
        # With ISIN letters are handled differently than for instance with
        # CUSIP, so we substitute them here
        $number =~ s/([A-Z])/$ctable{$1}/ge;
    }
    elsif ('imeisv' eq $self->{type}) {
        # With IMEISV the SV (software version) is left out from the
        # computation of the checkdigit
        $number = substr( $number, 0, 14 ) if ( 'imeisv' eq $self->{type} );
    }

    my @digits =  map { $ctable{$_} } split( //, $number );
    my $even   = 1;
    my $sum    = 0;
    for ( my $i = $#digits; $i >= 0; $i-- ) {
        if ($even) {
            my $tmp = 2 * $digits[$i];
            $sum += $tmp / 10 + $tmp % 10;
        }
        else {
            $sum += $digits[$i] / 10 + $digits[$i] % 10;
        }
        $even = not $even;
    }
    return ( 10 - $sum % 10 ) % 10;
}    # _compute_checkdigit()

sub _determine_pattern {
    my $self = shift;
    if ('cusip' eq $self->{type}) {
        $self->{pattern} = qr/[0-9A-Z]{8}/io;
    }
    else {
        $self->{pattern} = qr/[0-9A-Z ]+/io;
    }
} # _determine_pattern()

# Preloaded methods go here.

1;
__END__

=head1 NAME

CheckDigits::M10_001 - compute check digits for Bahncard (DE), IMEI,
IMEISV, ISIN, Miles&More, Payback (DE), Personnummer (SE), Passport
(BR), Credit Cards, SSN (US), Samordningsnummer (SE), VAT RN (ES), VAT
RN (IT), VAT RN (SE), International Securities Identifikation Number
(ISIN), CUSIP

=head1 SYNOPSIS

  use Algorithm::CheckDigits;

  $visa = CheckDigits('visa');

  if ($visa->is_valid('4111 1111 1111 1111')) {
	# do something
  }

  $cn = $visa->complete('4111 1111 1111 111');
  # $cn = '4111 1111 1111 1111'

  $cd = $visa->checkdigit('4111 1111 1111 1111');
  # $cd = '7'

  $bn = $visa->basenumber('4111 1111 1111 1111');
  # $bn = '4111 1111 1111 111'
  
=head1 DESCRIPTION

=head2 ALGORITHM

=over 4

=item 1

Beginning right all numbers are weighted alternatively 1 and 2 (that
is the check digit is weighted 1).

=item 2

The total of the digits of all products is computed.

=item 3

The sum of step 3 ist taken modulo 10.

=item 4

The check digit is the difference between 10 and the number from step
3.

=back

To validate the total of the digits of all numbers inclusive check
digit taken modulo 10 must be 0.

=head2 METHODS

=over 4

=item is_valid($number)

Returns true only if the last digit
is a valid check digit according to the algorithm given above.

Returns false otherwise,

If the checked number is of type CUSIP, the number must be exact 9 digits or
letters long and must not have spaces in between.

=item complete($number)

The check digit for C<$number> is computed and concatenated to the end
of C<$number>.

Returns the complete number with check digit or '' if C<$number>
does not consist solely of digits and spaces.

=item basenumber($number)

Returns the basenumber of C<$number> if C<$number> has a valid check
digit.

Return '' otherwise.

=item checkdigit($number)

Returns the checkdigit of C<$number> if C<$number> has a valid check
digit.

Return '' otherwise.

=back

=head2 EXPORT

None by default.

=head1 AUTHOR

Mathias Weidner, C<< <mamawe@cpan.org> >>

=head1 SEE ALSO

L<perl>,
L<Algorithm::CheckDigits>,
F<www.pruefziffernberechnung.de>.
F<http://en.wikipedia.org/wiki/CUSIP>

For IMEI, IMEISV: ETSI Technical Specification TS 100 508 (v6.2.0)

=cut
